File: electron/media-upload-recipes.js
/**
 * Provider-specific recipes for automated media upload via hidden BrowserWindow.
 * Each recipe defines selectors and behavior for DOM-based upload flows.
 * Selectors are candidate-based: try each until one matches. Fail fast on blocked indicators.
 *
 * Android vs Electron differences (documented to prevent drift):
 * - Android (MediaUploadRecipes.java): imgur/postimages only; no catbox (uses different flow).
 *   Uses WebChromeClient file chooser interception; trigger JS clicks file input.
 * - Electron: catbox/imgur/postimages. Uses CDP DOM.setFileInputFiles + submit button click.
 *   Catbox: Electron-only; keep behavior unchanged.
 * - Blocked/success selectors: reconciled with Android for imgur/postimages; Electron may add
 *   extra candidates (e.g. [data-captcha], .login-form) where DOM differs.
 * - Timeouts: imgur/postimages 45s (parity); catbox 30s (Electron-only).
 *
 * @typedef {Object} ProviderRecipe
 * @property {string} uploadUrl - Full URL of the provider's upload page
 * @property {readonly string[]} fileInputSelectorCandidates - CSS selectors for file input (first match wins)
 * @property {readonly string[]} submitSelectorCandidates - CSS selectors for submit/upload button
 * @property {Object} successExtractor - How to extract the result URL from the page
 * @property {readonly string[]} successExtractor.selectorCandidates - Selectors for element containing result URL
 * @property {'href'|'src'|'value'|'text'} successExtractor.attribute - Attribute or 'text' for textContent
 * @property {readonly string[]} blockedIndicators - Selectors that indicate captcha/login/challenge (fail immediately if present)
 * @property {number} timeoutMs - Max time to wait for upload completion
 */

/** @type {Readonly<Record<string, ProviderRecipe>>} */
export const MEDIA_UPLOAD_RECIPES = Object.freeze({
  catbox: Object.freeze({
    uploadUrl: 'https://catbox.moe',
    fileInputSelectorCandidates: Object.freeze(['input[type="file"]', 'input[type=file]']),
    submitSelectorCandidates: Object.freeze(['button[type="submit"]', 'input[type="submit"]', '[type="submit"]']),
    successExtractor: Object.freeze({
      selectorCandidates: Object.freeze(['a[href*="files.catbox.moe"]', 'input[value*="files.catbox.moe"]', '[class*="result"] a', 'textarea']),
      attribute: 'href',
    }),
    blockedIndicators: Object.freeze(['#challenge', '.captcha', '[data-captcha]', '.g-recaptcha', '.login-form', '#recaptcha']),
    timeoutMs: 30_000,
  }),
  /* Reconciled with Android MediaUploadRecipes (imgur): file input, success extractors, blocked indicators. */
  imgur: Object.freeze({
    uploadUrl: 'https://imgur.com/upload',
    fileInputSelectorCandidates: Object.freeze(['input[type="file"]', 'input[type=file]', '[data-file-input]']),
    submitSelectorCandidates: Object.freeze(['button[type="submit"]', '[data-action="upload"]', '.upload-btn', '[type="submit"]']),
    successExtractor: Object.freeze({
      selectorCandidates: Object.freeze(['a[href*="i.imgur.com"]', 'input[value*="i.imgur.com"]', '[class*="copy-link"] input', '[data-link]']),
      attribute: 'href',
    }),
    /* Android: .signin, .login. Electron adds: [data-captcha], .login-form for DOM variations. */
    blockedIndicators: Object.freeze(['#challenge', '.captcha', '[data-captcha]', '.g-recaptcha', '#recaptcha', '.login-form', '.signin', '.login']),
    timeoutMs: 45_000,
  }),
  /* Reconciled with Android MediaUploadRecipes (postimages): file input, success extractors, blocked indicators. */
  postimages: Object.freeze({
    uploadUrl: 'https://postimages.org',
    fileInputSelectorCandidates: Object.freeze(['input[type="file"]', 'input[type=file]', '#uploadFile', '.fileinput']),
    submitSelectorCandidates: Object.freeze(['button[type="submit"]', '[type="submit"]', '.btn-upload']),
    successExtractor: Object.freeze({
      selectorCandidates: Object.freeze(['input[readonly][value*="postimg"]', 'a[href*="i.postimg.cc"]', '[class*="direct-link"]', 'textarea']),
      attribute: 'value',
    }),
    blockedIndicators: Object.freeze(['#challenge', '.captcha', '.g-recaptcha', '#recaptcha']),
    timeoutMs: 45_000,
  }),
});

/**
 * Validates that every provider has required recipe fields: trigger (file input), success extractor,
 * blocked indicators, and timeout. Throws if any provider is invalid.
 */
function validateRecipes() {
  const required = ['fileInputSelectorCandidates', 'submitSelectorCandidates', 'successExtractor', 'blockedIndicators', 'timeoutMs'];
  for (const [provider, recipe] of Object.entries(MEDIA_UPLOAD_RECIPES)) {
    for (const key of required) {
      if (!(key in recipe) || recipe[key] == null) {
        throw new Error(`Recipe validation failed: ${provider} missing or null: ${key}`);
      }
    }
    const ex = recipe.successExtractor;
    if (!Array.isArray(ex?.selectorCandidates) || ex.selectorCandidates.length === 0 || !ex.attribute) {
      throw new Error(`Recipe validation failed: ${provider} successExtractor must have non-empty selectorCandidates and attribute`);
    }
    if (!Array.isArray(recipe.fileInputSelectorCandidates) || recipe.fileInputSelectorCandidates.length === 0) {
      throw new Error(`Recipe validation failed: ${provider} fileInputSelectorCandidates must be non-empty array`);
    }
    if (!Array.isArray(recipe.blockedIndicators) || recipe.blockedIndicators.length === 0) {
      throw new Error(`Recipe validation failed: ${provider} blockedIndicators must be non-empty array`);
    }
    if (typeof recipe.timeoutMs !== 'number' || recipe.timeoutMs <= 0) {
      throw new Error(`Recipe validation failed: ${provider} timeoutMs must be positive number`);
    }
  }
}

validateRecipes();

export { validateRecipes };

File: android/app/src/main/java/plebchan/android/MediaUploadRecipes.java
package fivechan.android;

/**
 * Provider-specific recipes for automated media upload via WebView.
 * Mirrors electron/media-upload-recipes.js for imgur/postimages.
 * Non-interactive automation; detects blocked states (captcha/login/challenge);
 * extracts direct media URL candidates via selector candidate arrays.
 */
public final class MediaUploadRecipes {

    public static final String PROVIDER_IMGUR = "imgur";
    public static final String PROVIDER_POSTIMAGES = "postimages";

    /** Max time to wait for upload completion (ms). */
    public static final long UPLOAD_TIMEOUT_MS = 45_000;
    /** Max time to wait for provider file input to be found/triggered (ms). */
    public static final long FILE_INPUT_TIMEOUT_MS = 8_000;
    /** Poll interval for success/blocked checks (ms). */
    public static final long POLL_INTERVAL_MS = 500;

    private MediaUploadRecipes() {}

    public static String getUploadUrl(String provider) {
        if (PROVIDER_IMGUR.equals(provider)) {
            return "https://imgur.com/upload";
        }
        if (PROVIDER_POSTIMAGES.equals(provider)) {
            return "https://postimages.org";
        }
        return null;
    }

    /**
     * JS to trigger file input click so WebChromeClient.onShowFileChooser fires.
     * Uses candidate selectors; first match wins.
     */
    public static String getTriggerFileInputJs(String provider) {
        String[] candidates;
        if (PROVIDER_IMGUR.equals(provider)) {
            candidates =
                    new String[] {
                        "input[type=\"file\"]",
                        "input[type=file]",
                        "[data-file-input]",
                    };
        } else if (PROVIDER_POSTIMAGES.equals(provider)) {
            candidates =
                    new String[] {
                        "input[type=\"file\"]",
                        "input[type=file]",
                        "#uploadFile",
                        ".fileinput",
                    };
        } else {
            return null;
        }
        return buildTriggerFileInputJs(candidates);
    }

    private static String buildTriggerFileInputJs(String[] selectors) {
        StringBuilder sb = new StringBuilder("(function(){var s=[");
        for (int i = 0; i < selectors.length; i++) {
            if (i > 0) sb.append(",");
            sb.append("\"").append(escapeJs(selectors[i])).append("\"");
        }
        sb.append("];for(var i=0;i<s.length;i++){var el=document.querySelector(s[i]);if(el){el.click();return true;}}return false;})()");
        return sb.toString();
    }

    /**
     * JS to extract direct media URL from page. Returns URL string or null.
     */
    public static String getSuccessJs(String provider) {
        String[] selectorCandidates;
        String attribute;
        if (PROVIDER_IMGUR.equals(provider)) {
            selectorCandidates =
                    new String[] {
                        "a[href*=\"i.imgur.com\"]",
                        "input[value*=\"i.imgur.com\"]",
                        "[class*=\"copy-link\"] input",
                        "[data-link]",
                    };
            attribute = "href";
        } else if (PROVIDER_POSTIMAGES.equals(provider)) {
            selectorCandidates =
                    new String[] {
                        "input[readonly][value*=\"postimg\"]",
                        "a[href*=\"i.postimg.cc\"]",
                        "[class*=\"direct-link\"]",
                        "textarea",
                    };
            attribute = "value";
        } else {
            return null;
        }
        return buildSuccessJs(selectorCandidates, attribute);
    }

    private static String buildSuccessJs(String[] selectors, String attr) {
        StringBuilder sb = new StringBuilder("(function(){var s=[");
        for (int i = 0; i < selectors.length; i++) {
            if (i > 0) sb.append(",");
            sb.append("\"").append(escapeJs(selectors[i])).append("\"");
        }
        sb.append("];for(var i=0;i<s.length;i++){try{var el=document.querySelector(s[i]);if(!el)continue;var v=(el.value||el.getAttribute(\"value\")||\"\").trim();var h=(el.getAttribute(\"href\")||el.href||\"\").trim();var t=(el.textContent||\"\").trim();if(h&&h.indexOf(\"http\")===0)return h;if(v&&v.indexOf(\"http\")===0)return v;if(t&&t.indexOf(\"http\")===0)return t}catch(e){}}return null})()");
        return sb.toString();
    }

    /**
     * JS to detect blocked state (captcha/login/challenge). Returns true if blocked.
     */
    public static String getBlockedJs(String provider) {
        String[] blockedIndicators;
        if (PROVIDER_IMGUR.equals(provider)) {
            blockedIndicators =
                    new String[] {
                        "#challenge", ".captcha", ".g-recaptcha", "#recaptcha", ".signin", ".login",
                    };
        } else if (PROVIDER_POSTIMAGES.equals(provider)) {
            blockedIndicators =
                    new String[] {
                        "#challenge", ".captcha", ".g-recaptcha", "#recaptcha",
                    };
        } else {
            return null;
        }
        return buildBlockedJs(blockedIndicators);
    }

    private static String buildBlockedJs(String[] selectors) {
        StringBuilder sb = new StringBuilder("(function(){var s=[");
        for (int i = 0; i < selectors.length; i++) {
            if (i > 0) sb.append(",");
            sb.append("\"").append(escapeJs(selectors[i])).append("\"");
        }
        sb.append("];for(var i=0;i<s.length;i++){if(document.querySelector(s[i]))return true}return false})()");
        return sb.toString();
    }

    private static String escapeJs(String s) {
        if (s == null) return "";
        return s.replace("\\", "\\\\").replace("\"", "\\\"");
    }
}
